Feasibility restoration problem for MDS system

nychiang opened the pull request:
This PR implements the feasibility restoration problem for MDS system.
It also:
1. implements the interface for warm-start.
2. fixes memory leak in the current unit test
3. implements missing functions in RAJA matrix kernel 
4. also closes #255

nychiang commented:
I will squash the commits before merging it.

nychiang commented:
@jwang125 I use `size_type` to replace `long long` in `hiopInterfacePrimalDecomp`.

nychiang commented:
I replaced `long long` by `index_type` and `size_type` in the tex files.
It is out of scope, but it's a quick fix for me. I search for the keyword `long long` and replace them one by one manually.

cnpetra commented:
> @jwang125 I use `size_type` to replace `long long` in `hiopInterfacePrimalDecomp`.

It's my doing as I forgot to replace those when I rebased Frank's branch

nychiang commented:
Fix the RAJA bug to access member variables. Thanks @ashermancinelli for all the valgrind testings and suggestions.
Now this branch passes all the CI tests.
Ready to review.

cnpetra commented on src/LinAlg/hiopMatrixRajaSparseTriplet.cpp:
```diff
@@ -1013,6 +1052,171 @@ hiopMatrixRajaSparseTriplet::RowStartsInfo::~RowStartsInfo()
+
+  assert(nnz_ == nnz_Jac_c_new + nnz_Jac_d_new);
+  
+  if(J_c.row_starts_host == nullptr){
```
space before bracket

cnpetra commented on src/LinAlg/hiopMatrixRajaSparseTriplet.cpp:
```diff
@@ -1013,6 +1052,171 @@ hiopMatrixRajaSparseTriplet::RowStartsInfo::~RowStartsInfo()
+  assert(J_c.row_starts_host);
+  int* Jc_row_st = J_c.row_starts_host->idx_start_;
+
+  if(J_d.row_starts_host == nullptr){
```
space before bracket

nychiang replied:
Fixed. I will wait for others' feedback before pushing the commit.

cnpetra commented on tests/LinAlg/matrixTestsDenseRowMajor.cpp:
```diff
@@ -156,6 +156,16 @@ const real_type* MatrixTestsDenseRowMajor::getLocalDataConst(hiop::hiopMatrixDen
+{
+  auto* amat = dynamic_cast<hiop::hiopMatrixDense*>(A);
+  if(amat == nullptr)
+    THROW_NULL_DEREF;
```
nothing bad with this, but we haven't use exceptions in the core of hiop. I would just return null_ptr...

nychiang replied:
I probably would keep this line, in order to keep consistence [with the other code in this file.](https://github.com/LLNL/hiop/blob/c559149f6cef1b71ebdea03cb56f820f29c2fde1/tests/LinAlg/matrixTestsRajaSparseTriplet.cpp)
`THROW_NULL_DEREF` is [defined here](https://github.com/LLNL/hiop/blob/c559149f6cef1b71ebdea03cb56f820f29c2fde1/tests/LinAlg/testBase.hpp#L58)

ashermancinelli replied:
We kept the exceptions out of hiop proper, but we found them to be really helpful when diagnosing bugs in our test suite and we probably won't miss the 2% performance hit. We can revert this, but I don't think we need to imo

cnpetra replied:
ok

cnpetra commented on tests/LinAlg/matrixTestsRajaDense.cpp:
```diff
@@ -162,6 +162,15 @@ const real_type* MatrixTestsRajaDense::getLocalDataConst(hiop::hiopMatrixDense*
+real_type* MatrixTestsRajaDense::getLocalData(hiop::hiopMatrixDense* A)
+{
+  auto* amat = dynamic_cast<hiop::hiopMatrixRajaDense*>(A);
+  if(amat == nullptr)
```
same here

cnpetra reviewed (changes requested):
lgtm... I left a couple of very small comments

pelesh commented on tests/LinAlg/matrixTestsSparse.hpp:
```diff
@@ -939,6 +939,118 @@ class MatrixTestsSparse : public TestBase
+    D.setToConstant(D_val);
+
+
+    const auto* C_iRow = getRowIndices(&C);
+    const auto* C_jCol = getColumnIndices(&C);
+    auto C_nnz = C.numberOfNonzeros();
+    const auto* D_iRow = getRowIndices(&D);
+    const auto* D_jCol = getColumnIndices(&D);
+    auto D_nnz = D.numberOfNonzeros();
```
I would really spell out data types here rather than using `auto`.

nychiang commented:
@cnpetra @pelesh I have addressed all the comments in the latest commit. PNNL CI is not working right now, but I think this easy change should be able to pass all the unit tests.
